File: verify/latency1r.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Daniel J. Dillberg
"""latency1r.py — LATENCY-1R (court B): render-start -> composited, the production render against the single-thread
reference, in the same sealed window session (off-gate, host).

    python verify/latency1r.py --host NAME [--session workshop/attest/sessionwalk-demo.json] [--per-cell 200] [--confirm]

LATENCY-1 cannot see the renderer (its instrument renders every frame before the window opens — LATENCY-1a). This
court, preregistered as its own rung before any host number, puts the render inside the clock: per sample the arm's
renderer runs on a pre-parsed sealed scene, the composite is blitted, and the sample ends when DwmFlush returns.
Two arms render the identical certified frame (PRODUCTION = fast::render at T=8; SINGLE-THREAD = the LOCKED
fast::emit), in two phase regimes (LOCKED: the render starts right after a composition; UNIFORM: at a seeded offset
uniform in [0, refresh) after one), interleaved ABBA. Witnesses come first, twice: the headless court
(`shell latency1r-selftest`, a deterministic mock surface) on THIS build, then the window court itself refuses to
take a number unless both arms reproduce every sealed frame witness and each other's composites.

Per regime the reading is computed from p50s (tail percentiles do not subtract): render delta dR = render(S) -
render(P), glass delta dG = total(S) - total(P), propagation = 1000*dG/dR permille (floored) — >= 500 PROPAGATES,
1..499 PARTIALLY ABSORBED, <= 0 ABSORBED, and dR <= 0 VOID (no render headroom in this apparatus). p99s are reported
beside it, never folded into it. Writes shell/attest/latency1r-<host>.json, citing LATENCY-1R. Never compared to
LATENCY-0/LATENCY-1 (a different observable) or to any emit p99; not input-to-photon.
"""
from __future__ import annotations

import argparse
import json
import os
import platform
import shutil
import subprocess
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "verify"))
import envelope  # noqa: E402

DEFAULT_SESSION = "workshop/attest/sessionwalk-demo.json"
PHASE_SEED = "0x5EED1A7E00000001"
PROD_THREADS = 8
PHASES = ("locked", "uniform")
ARMS = ("production", "single_thread")
SUBS = ("render_us", "present_us", "total_us")
PCT = ("p50", "p95", "p99", "max")
PROPAGATES_PERMILLE = 500


class Refuse(Exception):
    pass


def registry() -> dict:
    with open(os.path.join(ROOT, "verify", "preregister.json"), encoding="utf-8") as fh:
        return json.load(fh)["entries"]


def check_raw(raw: dict) -> dict:
    """The court's raw record, structurally: every cell, every sub-interval, integer percentiles, N per cell."""
    if raw.get("name") != "verdandi-latency1r":
        raise Refuse("not a verdandi-latency1r record")
    d = raw.get("data", {})
    if d.get("production_threads") != PROD_THREADS:
        raise Refuse(f"the production arm is not T={PROD_THREADS}")
    if d.get("phase_seed") != PHASE_SEED:
        raise Refuse("the phase seed is not the preregistered one")
    n = d.get("samples_per_cell")
    if not isinstance(n, int) or n <= 0 or not isinstance(d.get("refresh_period_us"), int) or d["refresh_period_us"] <= 0:
        raise Refuse("no samples per cell or no measured refresh")
    for ph in PHASES:
        for arm in ARMS:
            cell = d.get("cells", {}).get(ph, {}).get(arm)
            if not cell or cell.get("samples") != n:
                raise Refuse(f"cell {ph}/{arm} is missing or partial")
            for sub in SUBS:
                if not all(isinstance(cell.get(sub, {}).get(k), int) for k in PCT):
                    raise Refuse(f"cell {ph}/{arm} {sub} is not integer p50/p95/p99/max")
    return d


def classify_phase(cells: dict) -> dict:
    """The preregistered per-regime rule, on p50s: dR, dG, propagation permille, and the label."""
    P, S = cells["production"], cells["single_thread"]
    dR = S["render_us"]["p50"] - P["render_us"]["p50"]
    dG = S["total_us"]["p50"] - P["total_us"]["p50"]
    dPresent = P["present_us"]["p50"] - S["present_us"]["p50"]
    if dR <= 0:
        label, pm = "VOID", None
    else:
        pm = (1000 * dG) // dR
        label = "PROPAGATES" if pm >= PROPAGATES_PERMILLE else ("PARTIALLY ABSORBED" if pm > 0 else "ABSORBED")
    return {"label": label, "render_delta_p50_us": dR, "glass_delta_p50_us": dG, "propagation_permille": pm,
            "present_growth_p50_us": dPresent,
            "total_delta_p99_us": S["total_us"]["p99"] - P["total_us"]["p99"]}


def phase_sentence(ph: str, c: dict) -> str:
    if c["label"] == "VOID":
        return (f"{ph}: VOID — the production arm was not faster than the single-thread reference in the render "
                f"sub-interval (dR {c['render_delta_p50_us']} us), so there is no render headroom here to propagate.")
    s = (f"{ph}: {c['label']} — render delta {c['render_delta_p50_us']} us, glass delta {c['glass_delta_p50_us']} us, "
         f"{c['propagation_permille']} permille of the render headroom reached composited (p50s; the p99 total delta "
         f"is {c['total_delta_p99_us']} us, reported beside it).")
    if c["label"] != "PROPAGATES" and c["present_growth_p50_us"] > 0:
        s += (f" The production arm's frame-ready -> composited grew by {c['present_growth_p50_us']} us: the saved render "
              f"time was spent waiting for the composition.")
    return s


def seal_latency1r(raw: dict, reg: dict, host: str, session: dict, build: dict,
                   confirm_of: dict | None = None) -> tuple[dict, dict]:
    """Pure: the court's raw record -> the sealed LATENCY-1R record and the per-regime classification."""
    d = check_raw(raw)
    cls = {ph: classify_phase(d["cells"][ph]) for ph in PHASES}
    data = {
        "cells": {ph: {arm: {sub: {k: d["cells"][ph][arm][sub][k] for k in PCT} for sub in SUBS}
                       | {"samples": d["cells"][ph][arm]["samples"]} for arm in ARMS} for ph in PHASES},
        "refresh_period_us": d["refresh_period_us"],
        "sequence_frames": d["sequence_frames"],
        "samples_per_cell": d["samples_per_cell"],
        "production_threads": d["production_threads"],
        "phase_seed": d["phase_seed"],
        "derived": {ph: {k: v for k, v in cls[ph].items() if k != "label"} for ph in PHASES},
    }
    prov = {
        "tool": "shell latency1r-window (shell/latency1r.rs court over the GDI surface appended to shell/win32.rs) "
                "+ verify/latency1r.py",
        "surface": raw.get("provenance", {}).get("tool", ""),
        "host": host, "session": session, "build": build,
        "raw_unix_seconds": raw.get("provenance", {}).get("unix_seconds"),
        "python": platform.python_version(), "os": platform.system(), "machine": platform.machine(),
        "preregistered": {"rung": "LATENCY-1R", "chain_hash": reg["LATENCY-1R"]["chain_hash"]},
    }
    reading = ("LATENCY-1R, render-start -> composited, production (T=8) vs the single-thread reference, on host "
               f"{host} (refresh {d['refresh_period_us']} us, {d['samples_per_cell']} samples per cell). "
               + " ".join(phase_sentence(ph, cls[ph]) for ph in PHASES)
               + " A composed-GDI present is refresh-coupled (LATENCY-0), so absorption in the locked regime is a "
                 "property of this present path, not evidence that faster rendering is useless; decoupling it is "
                 "PRESENT-1's hypothesis. Within this apparatus only: never compared to LATENCY-0/LATENCY-1 or to any "
                 "emit p99; not input-to-photon.")
    name = "verdandi-latency1r"
    if confirm_of is not None:
        name = "verdandi-latency1r-confirm"
        prov["confirms"] = confirm_of
        orig = confirm_of.get("original_labels", {})
        same = all(orig.get(ph) == cls[ph]["label"] for ph in PHASES)
        was = ", ".join("%s %s" % (ph, orig.get(ph)) for ph in PHASES)
        now = ", ".join("%s %s" % (ph, cls[ph]["label"]) for ph in PHASES)
        reading = (f"CONFIRMATION run — does NOT replace the sealed LATENCY-1R record (chain {confirm_of['chain_hash'][:8]}). "
                   f"The per-regime shape {'REPRODUCED' if same else 'did NOT reproduce'}: original {was}; this run "
                   f"{now}. ") + reading
    rec = envelope.seal(
        name, 1, "measured", prov,
        {"certifies": f"render-start -> composited of the production render (T={PROD_THREADS}) and the single-thread "
                      f"reference over the sealed reference session on host {host}, locked and uniform phase, "
                      f"{d['samples_per_cell']} samples per cell over {d['sequence_frames']} frames"
                      + (" (a confirmation run)" if confirm_of is not None else ""),
         "host": host},
        ["input-to-photon latency: input transport and panel scan-out need capture hardware",
         "a comparison with LATENCY-0's or LATENCY-1's frame-ready -> composited, or with any emit p99 (different observables)",
         "a render-speed claim: that is GAUNTLET-2's; here the render is one part of a render-inclusive present interval",
         "that the locked or uniform regime is a real game loop under load (each is a declared phase model)",
         "that absorption in the locked regime means faster rendering is useless (it is the refresh-coupled GDI present's property)",
         "a single scalar: the propagation permille is on p50s; the p99s are reported beside it, never folded in",
         "any other host, session, panel or window state; absolute us are host- and panel-specific"],
        data, reading)
    return rec, {ph: cls[ph]["label"] for ph in PHASES}


def main() -> int:
    try:
        sys.stdout.reconfigure(errors="replace")
    except Exception:
        pass
    ap = argparse.ArgumentParser()
    ap.add_argument("--host", required=True)
    ap.add_argument("--session", default=DEFAULT_SESSION)
    ap.add_argument("--per-cell", type=int, default=200)
    ap.add_argument("--confirm", action="store_true",
                    help="a reproducibility run: seal a SEPARATE latency1r-confirm-<host>.json citing the sealed record")
    a = ap.parse_args()
    run = dict(capture_output=True, text=True, encoding="utf-8", errors="replace")
    try:
        if os.name != "nt":
            raise Refuse("LATENCY-1R is Windows-only (the instrument is the GDI/DWM window)")
        reg = registry()
        if "LATENCY-1R" not in reg:
            raise Refuse("LATENCY-1R is not preregistered — no number before the method")
        rustc = shutil.which("rustc")
        if rustc is None:
            raise Refuse("rustc not found")
        build_dir = os.path.join(ROOT, "verify", "build")
        os.makedirs(build_dir, exist_ok=True)
        exe = os.path.join(build_dir, "shell-latency1r.exe")
        flags = ["-O", "--cfg", "shell_window"]
        cp = subprocess.run([rustc] + flags + [os.path.join(ROOT, "shell", "main.rs"), "-o", exe], **run)
        if cp.returncode != 0:
            raise Refuse("rustc failed:\n" + cp.stderr)
        build = {"rustc": subprocess.run([rustc, "--version"], **run).stdout.strip(), "flags": flags}
        srec = envelope.read(os.path.join(ROOT, a.session))
        session = {"path": a.session, "chain_hash": srec["chain_hash"], "head": srec["data"]["head"]}
        # witnesses first, on THIS build: the headless court (mock surface) renders both arms to the sealed witnesses
        cp = subprocess.run([exe, "latency1r-selftest", "--session", a.session, "--per-cell", "1"], cwd=ROOT, **run)
        if cp.returncode != 0 or "latency1r court OK" not in cp.stdout:
            raise Refuse("the headless court did not reproduce the sealed witnesses — no number is taken\n" + cp.stdout + cp.stderr)
        raw_path = os.path.join(build_dir, f"latency1r-raw-{a.host}.json")
        if os.path.exists(raw_path):
            os.remove(raw_path)
        # the window court streams its progress and any refusal straight to this console (not captured)
        sys.stdout.flush()
        rc = subprocess.run([exe, "latency1r-window", "--session", a.session, "--per-cell", str(a.per_cell),
                             "--host", a.host, "--out", raw_path], cwd=ROOT).returncode
        if rc != 0 or not os.path.exists(raw_path):
            raise Refuse("the window court refused or wrote nothing (its reason is printed above)")
        with open(raw_path, encoding="utf-8") as fh:
            raw = json.load(fh)
        os.remove(raw_path)
        confirm_of = None
        name = f"latency1r-{a.host}.json"
        if a.confirm:
            canon = os.path.join(ROOT, "shell", "attest", name)
            if not os.path.exists(canon):
                raise Refuse(f"nothing to confirm — no sealed shell/attest/{name}; run without --confirm first")
            orig = envelope.read(canon)
            labels = {ph: classify_phase(orig["data"]["cells"][ph])["label"] for ph in PHASES}
            confirm_of = {"of": f"shell/attest/{name}", "chain_hash": orig["chain_hash"], "original_labels": labels}
            name = f"latency1r-confirm-{a.host}.json"
        rec, labels = seal_latency1r(raw, reg, a.host, session, build, confirm_of)
    except Refuse as e:
        print(f"REFUSE: {e}")
        return 2
    out_dir = os.path.join(ROOT, "shell", "attest")
    os.makedirs(out_dir, exist_ok=True)
    out = os.path.join(out_dir, name)
    envelope.write(out, rec)
    print("[latency1r] " + "; ".join(f"{ph} {labels[ph]}" for ph in PHASES))
    print(f"[latency1r] -> {os.path.relpath(out, ROOT)}  (cites LATENCY-1R {reg['LATENCY-1R']['chain_hash'][:8]})")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: shell/latency1r.rs
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Daniel J. Dillberg
//
// shell/latency1r.rs — LATENCY-1R: render-start -> composited, the production render against the single-thread
// reference, in the same sealed window session.
//
// LATENCY-0's instrument (and so LATENCY-1, which reuses it unchanged) times blit -> composited over frames that
// `playback::frames` rendered BEFORE the window opened, so the renderer is outside its clock. This court puts the
// render inside the clock: per sample, the arm's renderer runs on a pre-parsed sealed scene, the composite is
// converted and blitted, and the sample ends when the composition barrier returns. Two arms render the identical
// certified frame — Production (fast::render, T=8) and SingleThread (the LOCKED fast::emit) — and the render phase
// relative to composition is a declared variable with two regimes:
//
//     locked   the render starts immediately after a composition (a loop that renders right after it presents)
//     uniform  the render starts at a seeded pseudo-random offset, uniform in [0, refresh), after a composition
//
// The court is platform-agnostic: it runs against a `Surface` (a clock, a present, a composition barrier, a message
// pump). The Windows window is one Surface (appended to win32.rs, after LATENCY-0's untouched instrument, host-only);
// `MockSurface` is another, a deterministic clock the gate drives through `shell latency1r-selftest`, so everything
// here except the Win32 calls is exercised on every gate run. Witnesses come first: no clock starts until both arms
// have rendered every sealed frame to the sealed frame witness and to byte-identical composites.

use crate::mantle::{frame_digest, Scene};
use crate::present::{arm_composite, to_blit, Arm};

/// The preregistered phase seed (LATENCY-1R): the uniform regime's offsets are reproducible run to run.
pub const PHASE_SEED: u64 = 0x5EED_1A7E_0000_0001;
pub const PHASES: [&str; 2] = ["locked", "uniform"];
pub const ARMS: [Arm; 2] = [Arm::Production, Arm::SingleThread];

/// What the court needs from a window. `present` blits one frame and blocks on the composition barrier, returning
/// (frame-ready, composited) ticks exactly as LATENCY-0's present does: frame-ready is read right after the blit
/// call returns, composited right after the barrier returns.
pub trait Surface {
    fn ticks(&mut self) -> i64;
    fn freq(&self) -> i64;
    fn present(&mut self, bgr: &[u8]) -> Option<(i64, i64)>;
    fn flush(&mut self);
    fn pump(&mut self) -> bool;
}

/// One sealed move to render: its scene (parsed outside every clock) and its sealed frame witness.
pub struct FrameInput {
    pub scene: Scene,
    pub witness: String,
}

pub struct Cell {
    pub phase: usize,
    pub arm: Arm,
    pub render_us: Vec<u64>,
    pub present_us: Vec<u64>,
    pub total_us: Vec<u64>,
}

pub struct Court {
    pub refresh_us: u64,
    pub frames: usize,
    pub per_cell: usize,
    pub cells: Vec<Cell>,
}

fn splitmix(state: &mut u64) -> u64 {
    *state = state.wrapping_add(0x9E37_79B9_7F4A_7C15);
    let mut z = *state;
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    z ^ (z >> 31)
}

/// Nearest-rank percentiles, the same definition as LATENCY-0's (`win32.rs percentiles`).
pub fn percentiles(xs: &[u64]) -> (u64, u64, u64, u64) {
    let mut xs = xs.to_vec();
    xs.sort_unstable();
    let n = xs.len();
    let at = |p: f64| xs[(((n as f64) * p).ceil() as usize).saturating_sub(1).min(n - 1)];
    (at(0.50), at(0.95), at(0.99), xs[n - 1])
}

pub fn arm_name(arm: Arm) -> &'static str {
    match arm {
        Arm::Production => "production",
        Arm::SingleThread => "single_thread",
    }
}

/// The court. Witnesses first, then the refresh (LATENCY-0's estimate: the median of eight idle barrier intervals),
/// then `per_cell` rounds; each round renders the round's sealed frame once in each of the four cells
/// (locked/uniform x production/single_thread), in ABBA order (reversed on odd rounds) so drift and cache warmth
/// cancel between arms. Every sample starts from a composition. After each sample (outside its interval) the
/// composite is compared byte-for-byte to its pre-verified bytes. Any refusal returns Err and no record is written.
pub fn court(s: &mut dyn Surface, inputs: &[FrameInput], per_cell: usize) -> Result<Court, String> {
    if inputs.is_empty() || per_cell == 0 {
        return Err("LATENCY1R-EMPTY: no sealed frames or no samples requested".to_string());
    }
    // 1. witnesses first — outside every clock
    let mut expected: Vec<Vec<u8>> = Vec::with_capacity(inputs.len());
    for (i, f) in inputs.iter().enumerate() {
        let (fp, cp) = arm_composite(&f.scene, Arm::Production);
        if frame_digest(&fp) != f.witness {
            return Err(format!("LATENCY1R-WITNESS: the production arm's frame {} is not the sealed witness", i));
        }
        let (fs, cs) = arm_composite(&f.scene, Arm::SingleThread);
        if frame_digest(&fs) != f.witness {
            return Err(format!("LATENCY1R-WITNESS: the single-thread arm's frame {} is not the sealed witness", i));
        }
        if cp != cs {
            return Err(format!("LATENCY1R-WITNESS: the two arms' composites differ on frame {}", i));
        }
        expected.push(cp);
    }
    // 2. the refresh period — the median of eight idle composition intervals, LATENCY-0's estimate
    let freq = s.freq().max(1) as u128;
    let us = |d: i64| -> u64 { (d.max(0) as u128 * 1_000_000 / freq) as u64 };
    let mut idle: Vec<u64> = Vec::with_capacity(8);
    for _ in 0..8 {
        let a = s.ticks();
        s.flush();
        let b = s.ticks();
        idle.push(us(b - a));
    }
    idle.sort_unstable();
    let refresh_us = idle[idle.len() / 2];
    if refresh_us == 0 {
        return Err("LATENCY1R-NO-REFRESH: the composition barrier did not block".to_string());
    }
    // 3. the interleaved court
    let mut cells: Vec<Cell> = Vec::with_capacity(4);
    for c in 0..4 {
        cells.push(Cell { phase: c / 2, arm: ARMS[c % 2], render_us: Vec::with_capacity(per_cell),
                          present_us: Vec::with_capacity(per_cell), total_us: Vec::with_capacity(per_cell) });
    }
    let mut rng = PHASE_SEED;
    for round in 0..per_cell {
        let k = round % inputs.len();
        let order: [usize; 4] = if round % 2 == 0 { [0, 1, 2, 3] } else { [3, 2, 1, 0] };
        for &c in order.iter() {
            if !s.pump() {
                return Err("LATENCY1R-CLOSED: the window closed before the court finished; no partial record".to_string());
            }
            let (phase, arm) = (cells[c].phase, cells[c].arm);
            s.flush(); // the phase origin: every sample starts from a composition
            if phase == 1 {
                let offset_us = splitmix(&mut rng) % refresh_us;
                let wait = (offset_us as u128 * freq / 1_000_000) as i64;
                let target = s.ticks() + wait;
                while s.ticks() < target {}
            }
            let t0 = s.ticks();
            let (_frame, comp) = arm_composite(&inputs[k].scene, arm);
            let bgr = to_blit(&comp);
            let (t1, t2) = match s.present(&bgr) {
                Some(t) => t,
                None => return Err("LATENCY1R-NO-PRESENT: the blit or the composition barrier failed".to_string()),
            };
            if comp != expected[k] {
                return Err(format!("LATENCY1R-DRIFT: a composite of frame {} drifted from its verified bytes", k));
            }
            cells[c].render_us.push(us(t1 - t0));
            cells[c].present_us.push(us(t2 - t1));
            cells[c].total_us.push(us(t2 - t0));
        }
    }
    Ok(Court { refresh_us, frames: inputs.len(), per_cell, cells })
}

fn pct_json(xs: &[u64]) -> String {
    let (p50, p95, p99, max) = percentiles(xs);
    format!("{{\"p50\":{},\"p95\":{},\"p99\":{},\"max\":{}}}", p50, p95, p99, max)
}

fn esc(s: &str) -> String {
    let mut out = String::from("\"");
    for ch in s.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// The raw record the court's numbers go into. Sealing (the chain hash and the preregistration citation) is Python's
/// (`verify/latency1r.py`), so the envelope keeps one implementation. `data` holds numbers only — no verdict key.
pub fn raw_record(host: &str, surface: &str, c: &Court, unix_seconds: u64) -> String {
    let mut phases = Vec::new();
    for (p, pname) in PHASES.iter().enumerate() {
        let mut arms = Vec::new();
        for cell in c.cells.iter().filter(|x| x.phase == p) {
            arms.push(format!("\"{}\":{{\"render_us\":{},\"present_us\":{},\"total_us\":{},\"samples\":{}}}",
                              arm_name(cell.arm), pct_json(&cell.render_us), pct_json(&cell.present_us),
                              pct_json(&cell.total_us), cell.total_us.len()));
        }
        phases.push(format!("\"{}\":{{{}}}", pname, arms.join(",")));
    }
    let data = format!(
        "{{\"cells\":{{{}}},\"refresh_period_us\":{},\"sequence_frames\":{},\"samples_per_cell\":{},\"production_threads\":{},\"phase_seed\":\"0x{:016X}\"}}",
        phases.join(","), c.refresh_us, c.frames, c.per_cell, crate::fast::PROD_THREADS, PHASE_SEED);
    let prov = format!(
        "{{\"tool\":\"shell/latency1r.rs court over the {} surface\",\"host\":{},\"preregistered\":{{\"rung\":\"LATENCY-1R\",\"chain_hash\":\"PASTE_LATENCY1R_HASH\"}},\"unix_seconds\":{}}}",
        surface, esc(host), unix_seconds);
    let scope = format!("render-start -> composited of the production render (T={}) and the single-thread reference over the sealed reference session on host {}, locked and uniform phase, {} samples per cell over {} frames",
                        crate::fast::PROD_THREADS, host, c.per_cell, c.frames);
    format!(
        "{{\"name\":\"verdandi-latency1r\",\"version\":1,\"claim_class\":\"measured\",\"provenance\":{},\"validity_scope\":{{\"certifies\":{},\"host\":{}}},\"data\":{},\"reading\":\"render-start -> composited, production vs single-thread reference, per phase regime; NOT input-to-photon; never compared to LATENCY-0/LATENCY-1 or any emit p99\"}}",
        prov, esc(&scope), esc(host), data)
}

/// Human-readable lines (stdout) for the host operator and the gate.
pub fn summary(c: &Court) -> Vec<String> {
    let mut out = vec![format!("latency1r refresh_us {} frames {} per_cell {}", c.refresh_us, c.frames, c.per_cell)];
    for cell in &c.cells {
        let (r50, _, r99, _) = percentiles(&cell.render_us);
        let (p50, _, p99, _) = percentiles(&cell.present_us);
        let (t50, _, t99, tmax) = percentiles(&cell.total_us);
        out.push(format!("latency1r cell {} {} n={} render_p50={} render_p99={} present_p50={} present_p99={} total_p50={} total_p99={} total_max={}",
                         PHASES[cell.phase], arm_name(cell.arm), cell.total_us.len(), r50, r99, p50, p99, t50, t99, tmax));
    }
    out
}

/// A deterministic stand-in for the window so the gate drives the SAME court headless: a microsecond clock that
/// advances `step` per read, a composition every `period` µs (the barrier jumps to the next one), a present that
/// always lands, and a pump that reports the window closed after `close_after` pumps (a plant), if set.
pub struct MockSurface {
    pub t: i64,
    pub period: i64,
    pub step: i64,
    pub pumps: usize,
    pub close_after: Option<usize>,
}

impl MockSurface {
    pub fn new(period: i64, step: i64, close_after: Option<usize>) -> MockSurface {
        MockSurface { t: 0, period, step, pumps: 0, close_after }
    }
}

impl Surface for MockSurface {
    fn ticks(&mut self) -> i64 {
        self.t += self.step;
        self.t
    }
    fn freq(&self) -> i64 {
        1_000_000
    }
    fn present(&mut self, _bgr: &[u8]) -> Option<(i64, i64)> {
        let ready = self.ticks();
        self.flush();
        let composited = self.ticks();
        Some((ready, composited))
    }
    fn flush(&mut self) {
        self.t = (self.t / self.period + 1) * self.period;
    }
    fn pump(&mut self) -> bool {
        self.pumps += 1;
        match self.close_after {
            Some(k) => self.pumps <= k,
            None => true,
        }
    }
}
